### ISSUE TITLE
Flop reduction calculation

### ISSUE THREAD

**Issue by xuanlinli17** [yanghr/DeepHoyer#2]
Just wonder how is # flop reduction calculation calculated? Is there a script for it?

Thanks!

**Comment by yanghr** [comment]
For this project we haven't include an automated script for FLOPs computation. The way we get the results is by printing the non-zero dimensions with util.print_nonzeros (https://github.com/yanghr/DeepHoyer/blob/master/cifar/util.py#L26), then computes the FLOPs of each layer manually by multiplying the number of non-zero input channels, filters, times of convolution and the kernel size. Note that in some cases where the number of nonzero filters in layer i does not match the number of nonzero input channels in layer i+1's weight, we use the smaller one of the two for FLOPs computation, as there won't be computation needed if either the activation or the weight is all zero.

If you are interested in implementing a script for FLOPs computation, you may check this function (https://github.com/yanghr/SVD_Prune_EDLCV/blob/master/CNN/cifar10/resnet.py#L326) in my other repo. This script can accurately compute the FLOPs when the number of channels and filters are all matched in each pairs of consecutive layers.

**Comment by xuanlinli17** [comment]
Two questions:

1.

> Note that in some cases where the number of nonzero filters in layer i does not match the number of nonzero input channels in layer i+1's weight, we use the smaller one of the two for FLOPs computation 

For two consecutive conv layers, for example, the first one with filters [1,3,6] being zero and the second one with channels [1,2,4,7] being zero, according to the above statement, (#original_channel - 4) is used for flop computation. However, I think actually we can achieve an even smaller network by pruning the union of [1,3,6] and [1,2,4,7], so actually (#original_channel - 6) can be used for flop computation?

2. How is BatchNorm handled due to the bias term? Is it simply ignored for all channels that equal zero?

**Comment by yanghr** [comment]
For your first question, you are absolutely correct on this, achieving a smaller architecture is possible if we take a closer look at the pruned structure. In this DeepHoyer paper I didn't consider this part, and only reports the "worst" possible #FLOPs based on the dimension of each channel's weight. So carefully evaluating the pruned model to figure out whether more filters can be removed, or adding some constraints in the training process to prevent these mismatch in filters to happen could potentially further boost the results.

For your second question, yes the batchnorm is ignored. Also typically we don't consider the computation within batchnorm into the total FLOPs of the model, as they are not performed in a multiplication-accumulation way as in Conv and FC layers, and the computation amount is neglectable comparing to the rest of the model.

**Comment by huangbiubiu** [comment]
This function might help you to compute the FLOPs automatically. Hope this could help you.

```python3
def compute_conv_flops(model, cuda=False, prune=False):
    """
    compute the FLOPs for sparse model
    @param cuda: if use gpu to do the forward pass
    @param prune: if compute the FLOPs of the pruned model. If prune == False, the method compute the baseline FLOPs.
    """
    if isinstance(model, torch.nn.DataParallel):
        model = model.module

    list_conv = []
    list_linear = []

    def conv_hook(self, input, output):

        batch_size, input_channels, input_height, input_width = input[0].size()

        # prune the conv layers
        tensor = self.weight.data.cpu().numpy()
        tensor = np.abs(tensor)
        if len(tensor.shape) == 4:
            dim0 = np.sum(np.sum(tensor, axis=0), axis=(1, 2))
            dim1 = np.sum(np.sum(tensor, axis=1), axis=(1, 2))
        if len(tensor.shape) == 2:
            dim0 = np.sum(tensor, axis=0)
            dim1 = np.sum(tensor, axis=1)
        nz_count0 = np.count_nonzero(dim0)  # input channel
        nz_count1 = np.count_nonzero(dim1)  # output channel

        output_channels, output_height, output_width = output[0].size()

        if prune:
            input_channels = nz_count0
            output_channels = nz_count1

        kernel_ops = self.kernel_size[0] * self.kernel_size[1] * (input_channels / self.groups)

        flops = kernel_ops * output_channels * output_height * output_width

        list_conv.append(flops)

    def linear_hook(self, input, output):
        weight_ops = self.weight.nelement()

        flops = weight_ops
        list_linear.append(flops)

    def add_hooks(net, hook_handles: list):
        """
        apply FLOPs handles to conv layers recursively
        """
        children = list(net.children())
        if not children:
            if isinstance(net, torch.nn.Conv2d):
                hook_handles.append(net.register_forward_hook(conv_hook))
            if isinstance(net, torch.nn.Linear):
                hook_handles.append(net.register_forward_hook(linear_hook))
            return
        for c in children:
            add_hooks(c, hook_handles)

    handles = []
    add_hooks(model, handles)
    demo_input = torch.rand(8, 3, 224, 224)
    if cuda:
        demo_input = demo_input.cuda()
        model = model.cuda()
    model(demo_input)

    total_flops = sum(list_conv) + sum(list_linear)

    # clear handles
    for h in handles:
        h.remove()
    return total_flops


```
